compiler/semantic/sql.go

```diff
@@ -322,35 +325,141 @@ func (a *analyzer) semSelectValue(sel *ast.Select, sch schema, seq dag.Seq) (dag
+			e = &dag.RecordExpr{Kind: "RecordExpr", Elems: []dag.RecordElem{f}}
+			val, _ := kernel.EvalAtCompileTime(sctx, e)
+			recType = super.TypeUnder(val.Type()).(*super.TypeRecord)
+		}
+		if schema == nil {
+			schema = recType
+		} else if schema != recType {
+			a.error(astExpr, errors.New("values clause must contain uniformly typed values"))
+		}
+		exprs = append(exprs, e)
+	}
+	if schema == nil {
+		a.error(values, errors.New("values clause must contain uniformly typed values"))
+		return seq, badSchema()
+	}
+	columns := make([]string, 0, len(schema.Fields))
+	for _, f := range schema.Fields {
+		columns = append(columns, f.Name)
+	}
+	seq = append(seq, &dag.Yield{
+		Kind:  "Yield",
+		Exprs: exprs,
+	})
+	return seq, &anonSchema{columns}
+}
+
 func (a *analyzer) genDistinct(e dag.Expr, seq dag.Seq) dag.Seq {
 	return append(seq, &dag.Distinct{
 		Kind: "Distinct",
 		Expr: e,
 	})
 }
 
-func (a *analyzer) semSQLPipe(op *ast.SQLPipe, seq dag.Seq, alias string) (dag.Seq, schema) {
+func (a *analyzer) semSQLPipe(op *ast.SQLPipe, seq dag.Seq, alias *ast.TableAlias) (dag.Seq, schema) {
 	if len(op.Ops) == 1 && isSQLOp(op.Ops[0]) {
 		seq, sch := a.semSQLOp(op.Ops[0], seq)
-		return derefSchema(sch, alias, seq)
+		outSeq, outSch, err := derefSchemaWithAlias(sch, alias, seq)
+		if err != nil {
+			a.error(op.Ops[0], err)
+		}
+		return outSeq, outSch
 	}
 	if len(seq) > 0 {
 		panic("semSQLOp: SQL pipes can't have parents")
 	}
-	return a.semSeq(op.Ops), &dynamicSchema{name: alias}
+	var name string
+	if alias != nil {
+		name = alias.Name
+		if len(alias.Columns) != 0 {
+			a.error(alias, errors.New("cannot apply aliased columns to dynamically typed data"))
+		}
+	}
+	return a.semSeq(op.Ops), &dynamicSchema{name: name}
 }
 
-func derefSchema(sch schema, alias string, seq dag.Seq) (dag.Seq, schema) {
-	e, sch := sch.deref(alias)
+func derefSchemaAs(sch schema, table string, seq dag.Seq) (dag.Seq, schema) {
+	e, sch := sch.deref(table)
 	if e != nil {
 		seq = yieldExpr(e, seq)
 	}
 	return seq, sch
 }
 
+func derefSchema(sch schema, seq dag.Seq) (dag.Seq, schema) {
+	return derefSchemaAs(sch, "", seq)
+}
+
+func derefSchemaWithAlias(insch schema, alias *ast.TableAlias, inseq dag.Seq) (dag.Seq, schema, error) {
+	var table string
+	if alias != nil {
+		table = alias.Name
+	}
+	seq, sch := derefSchemaAs(insch, table, inseq)
+	if alias == nil || len(alias.Columns) == 0 {
+		return seq, sch, nil
+	}
+	switch sch := sch.(type) {
+	case *anonSchema:
+		return mapColumns(sch.columns, alias, seq)
+	case *staticSchema:
+		return mapColumns(sch.columns, alias, seq)
+	default:
+		return seq, sch, errors.New("cannot apply column aliases to dynamically typed data")
+	}
+}
+
+func mapColumns(in []string, alias *ast.TableAlias, seq dag.Seq) (dag.Seq, schema, error) {
+	if len(alias.Columns) > len(in) {
+		return nil, nil, fmt.Errorf("cannot apply %d aliased columns in alias %q to table with %d columns", len(alias.Columns), alias.Name, len(in))
```

Comment:
Nit: Maybe clearer like this?
```suggestion
		return nil, nil, fmt.Errorf("cannot apply %d column aliases to table %q with %d columns", len(alias.Columns), alias.Name, len(in))
```